applications/mne_sample_set_downloader/downloader.cpp

```diff
@@ -0,0 +1,216 @@
+    ui->filePath->setText(QDir::toNativeSeparators(dir));
+}
+
+void Downloader::readyToExtract(bool stat, QString zip, QString current)
+{
+    if (stat == true)
+    {
+        ui->label->setText("EXTRACTING...");
+        ui->progressBar->setVisible(false);
+        QObject::connect(&m_extractor, SIGNAL(zipperError()), this, SLOT(checkZipperPath()));
+        connect(&m_extractor, SIGNAL(extractionDone()), this, SLOT(done()));
+        m_extractor.beginExtraction(zip, current);
+    }
+}
+
+void Downloader::downloadFinished()
+{
+    if (m_qReply != NULL)
+        m_qFile.write(m_qReply->readAll());
+    m_qFile.close();
+    boolDownloadStatus = true;
+    QString zipPath ="\"" + QString (ui->filePath->text()) + "\"";
+    ui->progressBar->setEnabled(false);
+    readyToExtract(boolDownloadStatus, zipPath, m_qCurrentPath);
+    return;
+}
+#elif __linux__
+
+void Downloader::readyToExtract(bool stat)
+{
+    if (stat == true)
+    {
+        ui->label->setText("EXTRACTING...");
+        m_extractor.beginExtraction();
+    }
+}
+
+void Downloader::downloadFinished()
+{
+    if (m_qReply != NULL)
+        m_qFile.write(m_qReply->readAll());
+    boolDownloadStatus = true;
+    ui->progressBar->setEnabled(false);
+    readyToExtract(boolDownloadStatus);
+    return;
+}
+
+#endif
+
+void Downloader::done()
+{
+    ui->label->setText("done");
+}
+
+//placeholder downloadFinished OSX
+
+void Downloader::dataReady()
+{
+    if (m_qReply != NULL)
+        m_qFile.write(m_qReply->readAll());
+    return;
+}
+
+void Downloader::downloadProgress(qint64 recieved,qint64 total)
+{
+    ui->progressBar->setMaximum(total);
+    ui->progressBar->setValue(recieved);
+}
+
+
+void Downloader::on_downloadButton_clicked()
+{
+    ui->downloadButton->setEnabled(false);
+    ui->progressBar->show();
+    ui->label->setText("DOWNLOADING FILES...");
+
+    QNetworkAccessManager *manager = new QNetworkAccessManager;
+    QString url = "ftp://surfer.nmr.mgh.harvard.edu/pub/data/MNE-sample-data-processed.tar.gz"; //"ftp://surfer.nmr.mgh.harvard.edu/pub/data/test_data.tar.gz";
+    m_qReply = QSharedPointer<QNetworkReply> (manager->get(QNetworkRequest(url)));
+
+    connect(m_qReply.data(),SIGNAL(readyRead()),this,SLOT(dataReady()));
+    connect(m_qReply.data(),SIGNAL(finished()), this,SLOT(downloadFinished()));
+
+    if (m_qFile.fileName() != "sample.tar.gz")
+        m_qFile.setFileName("sample.tar.gz");
+    m_qFile.open(QIODevice::Append);
+
+    connect(m_qReply.data(), SIGNAL(downloadProgress(qint64,qint64)), this, SLOT(downloadProgress(qint64,qint64)));
+    QEventLoop loop;
+    connect(m_qReply.data(), SIGNAL(finished()), &loop, SLOT(quit()));
+    loop.exec();
+    return;
+}
+
+void Downloader::on_exitButton_clicked()
+{
+    this->close();
+    m_qFile.remove();
+    m_qFile.close();
+    return;
+}
+
+
+Downloader::Downloader(QWidget *parent) :
```

Comment:
Please keep the same order of the functions in the cpp file as in the header file